![](img/shark.png)

# Global Shark Attack

Fecha de entrega: _23:59 05/04/2021_ 

### ¿En qué consiste este mini-proyecto?

El objetivo de este mini-proyecto es que os enfrentéis a realizar un Exploratory Data Analysis por primera vez, antes del proyecto EDA. 

Durante este mini-proyecto deberás poner en práctica todo lo aprendido durante las últimas semanas. Tanto cómo te haga falta y solo si te hace falta. Esta vez serás tú el qué tome la decisión de qué tratamientos o limpiezas y visualizaciones serán mejor para tratar tus datos. 

Mínimo deberás importar las siguientes librerías, aunque te animo a explorar otras librerías nuevas.

```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
```

Todo esto con el fin de poder verificar nuestra hipótesis. Sigue leyendo para más información.

### Dataset

El dataset que utilizaremos para este mini-proyecto contiene datos sobre los ataques de tiburones en el mundo: localicación, fechas, tipología, especie causante, información sobre las víctimas, etc.. 

Registrado por http://www.sharkattackfile.net/index.htm.

Te lo podrás descargar en este enlace: https://www.kaggle.com/teajay/global-shark-attacks

### Hipótesis
Elige una o varias de estas hipótesis para desarrollar tu mini proyecto.

Siéntete libre de establecer nuevas hipótesis que puedas validar o rechazar con tu análisis.

- El número de supervivientes ha aumentado con el paso de los años
- Todos los ataques producidos en la India, ocurren a primera hora de la mañana
- La mayoría de los ataques en los últimos 200 años tuvieron lugar en USA, Australia y Sudáfrica. 
- Las actividades con más riesgo de sufrir un ataque de tiburon son pescar y nadar. 
- La mayoría de los ataques de tiburón fueron fatales.
- Los tiburones tienen especial predilección por los surferos.
- Reunion Island es el lugar del mundo donde ocurren más ataques de tiburón en terminos relativos.
- En Sudáfrica hay más ataques letales a mujeres que en Australia.

### Github*
*_Este punto es opcional pero MUY recomendable, ya que en el Proyecto de EDA será obligatorio_

Crea un repositorio nuevo para guardar los archivos que necesites.

Sigue la siguiente estructuras de carpetas:

    - src -> aquí encontrarás el código, en este caso los notebooks que uses
    - input -> aquí guardarás el dataset que utilices
    - output -> aquí guardarás gráficas o datasets limpios, si los tuvieras

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Las hipótesis que se van a tratar de refutar/validar son las siguientes:
* Todos los ataques producidos en la India, ocurren a primera hora de la mañana.
* Las actividades con más riesgo de sufrir un ataque de tiburón son pescar o nadar
* Reunion Island es el lugar del mundo donde ocurren más ataques de tiburón en terminos relativos.

In [2]:
sharks = pd.read_csv('../input/attacks.csv',encoding = "ISO-8859-1")
sharks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [3]:
# Con qué datos contamos??
sharks.describe()

Year  original order
count  6300.000000     6309.000000
mean   1927.272381     3155.999683
std     281.116308     1821.396206
min       0.000000        2.000000
25%    1942.000000     1579.000000
50%    1977.000000     3156.000000
75%    2005.000000     4733.000000
max    2018.000000     6310.000000

In [4]:
print("DataFrame shape:", sharks.shape, '\n')
print("DataFrame info:",sharks.info(), '\n')
print("DataFrame Columns:", sharks.columns, '\n')
print("type of each column:", sharks.dtypes, '\n')
print("Missing values in every colum", sharks.isna().sum())

DataFrame shape: (25723, 24) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species

Puede verse en los resultados de la celda anteriro que hay bastantes nulos en el DataFrame, así que, antes de hacer nada, lo primero es limpiarlo

## 1. Limpieza del DataFrame
Observando los resultados, se vaa a proceder de la siguiente manera:
* Las columnas `Unnamed: 22` y `Unnamed: 23` se eliminan completamente del data set por no aportar valor ninguno.
* También hay columnas como `Age`, `Time`, `Species` que tienen bastantes nulos (comparándolo con el resto de columnas)
* Se eliminarán los registros con nulos a lo largo de toda la fila (si exiten)
* Se comprobará que no hay duplicados

In [5]:
#Eliminar las dos últimas columnas
sharks = sharks.loc[:, 'Date': 'original order']
sharks.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

   ... Fatal (Y/N)           Time         Species   \
0  ...           N          18h00      White shark   
1  ...           N  14h00  -15h00              NaN   
2  ...           N          07h45              NaN   
3  ...           N            NaN        2 m shark   
4  ...           N            NaN  Tiger shark, 3m   

           Investigator or Source                       pdf  \
0                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order  
0    2018.06.25         6303.0  
1    2018.06.18         6302.0  
2    2018.06.09         6301.0  
3    2018.06.08         6300.0  
4    2018.06.04         6299.0  

[5 rows x 21 columns]

In [6]:
#Limpieza de los registros sin datos
print("Antes de limpiar los registros sin datos el tamaño del DataFrame es:",sharks.shape, "\n")
print("Después de limpiar los registros el tamaño del Dataframe es:",sharks.dropna(how='all').shape, "\n")
print('Se ha eliminado', sharks.shape[0] - sharks.dropna(how='all').shape[0], 'filas')

Antes de limpiar los registros sin datos el tamaño del DataFrame es: (25723, 21) 

Después de limpiar los registros el tamaño del Dataframe es: (6309, 21) 

Se ha eliminado 19414 filas


Como se puede ver, se han eliminado 19414 filas que no contenían datos, por ello, se procede a realizar un cambio permanente con 'inplace', y se guardará ese dataframe en un nuevo csv para poder realizar tratamientos posteriores. Pero antes, continuamos con la limpieza.

In [7]:
df_sharks = sharks.dropna(how='all')

In [8]:
df_sharks.tail(20)

Date  Year        Type  \
6289  No date (3 days after preceding incident) & pr...   0.0  Unprovoked   
6290                                 Before 19-Jul-1913   0.0  Unprovoked   
6291                                        Before 1911   0.0  Unprovoked   
6292                                         Circa 1862   0.0  Unprovoked   
6293                                        Before 1906   0.0  Unprovoked   
6294                                        Before 1906   0.0  Unprovoked   
6295                                        Before 1906   0.0  Unprovoked   
6296                                        Before 1906   0.0  Unprovoked   
6297                                        Before 1903   0.0  Unprovoked   
6298                                        Before 1903   0.0  Unprovoked   
6299                                          1900-1905   0.0  Unprovoked   
6300                                          1883-1889   0.0  Unprovoked   
6301                                          1845-1853   0.0  Unprovoked   
6302                                                NaN   NaN         NaN   
6303                                                NaN   NaN         NaN   
6304                                                NaN   NaN         NaN   
6305                                                NaN   NaN         NaN   
6306                                                NaN   NaN         NaN   
6307                                                NaN   NaN         NaN   
6308                                                NaN   NaN         NaN   

                 Country               Area  \
6289        SOUTH AFRICA      KwaZulu-Natal   
6290        SOUTH AFRICA      KwaZulu-Natal   
6291               ASIA?                NaN   
6292                 USA             Hawaii   
6293           AUSTRALIA                NaN   
6294           AUSTRALIA                NaN   
6295           AUSTRALIA                NaN   
6296           AUSTRALIA    New South Wales   
6297           AUSTRALIA  Western Australia   
6298           AUSTRALIA  Western Australia   
6299                 USA     North Carolina   
6300              PANAMA                NaN   
6301  CEYLON (SRI LANKA)   Eastern Province   
6302                 NaN                NaN   
6303                 NaN                NaN   
6304                 NaN                NaN   
6305                 NaN                NaN   
6306                 NaN                NaN   
6307                 NaN                NaN   
6308                 NaN                NaN   

                                 Location      Activity  \
6289                               Durban       Fishing   
6290                               Durban        Wading   
6291                                  NaN      Swimming   
6292                                 Puna           NaN   
6293                                  NaN       Fishing   
6294                                  NaN       Fishing   
6295                                  NaN       Fishing   
6296                                           Swimming   
6297                          Roebuck Bay        Diving   
6298                                  NaN  Pearl diving   
6299                       Ocracoke Inlet      Swimming   
6300                 Panama Bay 8ºN, 79ºW           NaN   
6301  Below the English fort, Trincomalee      Swimming   
6302                                  NaN           NaN   
6303                                  NaN           NaN   
6304                                  NaN           NaN   
6305                                  NaN           NaN   
6306                                  NaN           NaN   
6307                                  NaN           NaN   
6308                                  NaN           NaN   

                       Name Sex   Age  ... Fatal (Y/N) Time  \
6289     a native fisherman    M  NaN  ...           Y  NaN   
6290       a young Scotsman    M  NaN  ...           Y  NaN   
6291            

Viendo el DataFrame como está, se puede ver que una gran parte de las últimas filas (7) contienen muchos nulos.

Por eso aplico un 'thresh' a este DataFrame.

Existen 21 columnas, de manera que eliminaré los registros que solo tengan 2 valores no nulos (algo menos de un 10% de valores no nulos en cada registro, si no, no lo daré por válido).

In [9]:
clean_sharks = df_sharks.dropna(thresh=19)

In [10]:
clean_sharks

Date    Year        Type             Country              Area  \
0     25-Jun-2018  2018.0     Boating                 USA        California   
1     18-Jun-2018  2018.0  Unprovoked                 USA           Georgia   
2     09-Jun-2018  2018.0     Invalid                 USA            Hawaii   
3     08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   New South Wales   
4     04-Jun-2018  2018.0    Provoked              MEXICO            Colima   
...           ...     ...         ...                 ...               ...   
6265    1941-1942     0.0  Unprovoked                IRAQ            Basrah   
6266  1940 - 1950     0.0  Unprovoked        SAUDI ARABIA  Eastern Province   
6267  1940 - 1950     0.0  Unprovoked        SAUDI ARABIA  Eastern Province   
6296  Before 1906     0.0  Unprovoked           AUSTRALIA   New South Wales   
6301    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)  Eastern Province   

                                         Location     Activity  \
0                     Oceanside, San Diego County     Paddling   
1                  St. Simon Island, Glynn County     Standing   
2                                    Habush, Oahu      Surfing   
3                              Arrawarra Headland      Surfing   
4                                        La Ticla  Free diving   
...                                           ...          ...   
6265  Shatt-el Arab River near a small boat stand     Swimming   
6266          East of the Ras Tanura-Jubail area        Diving   
6267          East of the Ras Tanura-Jubail area        Diving   
6296                                                  Swimming   
6301          Below the English fort, Trincomalee     Swimming   

                     Name Sex        Age  ... Fatal (Y/N)           Time  \
0             Julie Wolfe    F        57  ...           N          18h00   
1         Adyson McNeely     F        11  ...           N  14h00  -15h00   
2             John Denges    M        48  ...           N          07h45   
3                    male    M       NaN  ...           N            NaN   
4          Gustavo Ramos     M       NaN  ...           N            NaN   
...                   ...  ...       ...  ...         ...            ...   
6265                 male    M  13 or 14  ...           Y      Afternoon   
6266        a pearl diver    M       NaN  ...           Y            NaN   
6267  a fisherman / diver    M       NaN  ...           N            NaN   
6296             Arab boy    M       NaN  ...           Y            NaN   
6301                 male    M        15  ...           Y            NaN   

                                               Species   \
0                                           White shark   
1                                                   NaN   
2                                                   NaN   
3                                             2 m shark   
4                                       Tiger shark, 3m   
...                                                 ...   
6265                                         Bull shark   
6266                             "a black-tipped shark"   
6267                                           6' shark   
6296  Said to involve a grey nurse shark that leapt ...   
6301                                                NaN   

                                 Investigator or Source  \
0                                      R. Collier, GSAF   
1                        K.McMurray, TrackingSharks.com   
2                        K.McMurray, TrackingSharks.com   
3                                        B. Myatt, GSAF   
4                                             A .Kipper   
...                                                 ...   
6265                       B.W. Coad & L.A.J. Al-Hassan   
6266                                          G.F. Mead   
6267                                          G.F. Mead   
6296  L. Becke in New York Sun, 9/9/1906; L. Schultz...   


In [11]:
#Aplicando mismo % de no nulos a las columnas
clean_sharks.dropna(axis=1, thresh=400)

Date    Year        Type             Country              Area  \
0     25-Jun-2018  2018.0     Boating                 USA        California   
1     18-Jun-2018  2018.0  Unprovoked                 USA           Georgia   
2     09-Jun-2018  2018.0     Invalid                 USA            Hawaii   
3     08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   New South Wales   
4     04-Jun-2018  2018.0    Provoked              MEXICO            Colima   
...           ...     ...         ...                 ...               ...   
6265    1941-1942     0.0  Unprovoked                IRAQ            Basrah   
6266  1940 - 1950     0.0  Unprovoked        SAUDI ARABIA  Eastern Province   
6267  1940 - 1950     0.0  Unprovoked        SAUDI ARABIA  Eastern Province   
6296  Before 1906     0.0  Unprovoked           AUSTRALIA   New South Wales   
6301    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)  Eastern Province   

                                         Location     Activity  \
0                     Oceanside, San Diego County     Paddling   
1                  St. Simon Island, Glynn County     Standing   
2                                    Habush, Oahu      Surfing   
3                              Arrawarra Headland      Surfing   
4                                        La Ticla  Free diving   
...                                           ...          ...   
6265  Shatt-el Arab River near a small boat stand     Swimming   
6266          East of the Ras Tanura-Jubail area        Diving   
6267          East of the Ras Tanura-Jubail area        Diving   
6296                                                  Swimming   
6301          Below the English fort, Trincomalee     Swimming   

                     Name Sex        Age  ... Fatal (Y/N)           Time  \
0             Julie Wolfe    F        57  ...           N          18h00   
1         Adyson McNeely     F        11  ...           N  14h00  -15h00   
2             John Denges    M        48  ...           N          07h45   
3                    male    M       NaN  ...           N            NaN   
4          Gustavo Ramos     M       NaN  ...           N            NaN   
...                   ...  ...       ...  ...         ...            ...   
6265                 male    M  13 or 14  ...           Y      Afternoon   
6266        a pearl diver    M       NaN  ...           Y            NaN   
6267  a fisherman / diver    M       NaN  ...           N            NaN   
6296             Arab boy    M       NaN  ...           Y            NaN   
6301                 male    M        15  ...           Y            NaN   

                                               Species   \
0                                           White shark   
1                                                   NaN   
2                                                   NaN   
3                                             2 m shark   
4                                       Tiger shark, 3m   
...                                                 ...   
6265                                         Bull shark   
6266                             "a black-tipped shark"   
6267                                           6' shark   
6296  Said to involve a grey nurse shark that leapt ...   
6301                                                NaN   

                                 Investigator or Source  \
0                                      R. Collier, GSAF   
1                        K.McMurray, TrackingSharks.com   
2                        K.McMurray, TrackingSharks.com   
3                                        B. Myatt, GSAF   
4                                             A .Kipper   
...                                                 ...   
6265                       B.W. Coad & L.A.J. Al-Hassan   
6266                                          G.F. Mead   
6267                                          G.F. Mead   
6296  L. Becke in New York Sun, 9/9/1906; L. Schultz...   


Antes de guardarlo en un csv, vulvo a analizar los datos:

In [12]:
print("DataFrame shape:", clean_sharks.shape, '\n')
print("DataFrame info:",clean_sharks.info(), '\n')
print("DataFrame Columns:", clean_sharks.columns, '\n')
print("type of each column:", clean_sharks.dtypes, '\n')
print("Missing values in every colum", clean_sharks.isna().sum())

DataFrame shape: (4138, 21) 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4138 entries, 0 to 6301
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Date                    4138 non-null   object 
 1   Year                    4138 non-null   float64
 2   Type                    4138 non-null   object 
 3   Country                 4138 non-null   object 
 4   Area                    4092 non-null   object 
 5   Location                4089 non-null   object 
 6   Activity                4083 non-null   object 
 7   Name                    4133 non-null   object 
 8   Sex                     4078 non-null   object 
 9   Age                     3261 non-null   object 
 10  Injury                  4137 non-null   object 
 11  Fatal (Y/N)             3926 non-null   object 
 12  Time                    2840 non-null   object 
 13  Species                 2828 non-null   object 
 14  Investigat

Finalizo este apartado creando un nuevo csv que utilizaré para el análisis.

In [13]:
clean_sharks.loc[:,['Activity']].head(20)

Activity
0          Paddling
1          Standing
2           Surfing
3           Surfing
4       Free diving
6          Swimming
7           Fishing
8           Walking
9          Standing
10   Feeding sharks
11  Boogie boarding
12         Swimming
13          Fishing
14   Feeding sharks
15          Surfing
16          Surfing
18     Scuba diving
19          Surfing
20          Surfing
21          Surfing

## 2. Análisis de los datos

Como mencioné en el [*apartado 01*](#1.-Limpieza-del-DataFrame) son las siguientes:

* Todos los ataques producidos en la India, ocurren a primera hora de la mañana.
* Las actividades con más riesgo de sufrir un ataque de tiburón son pescar o nadar
* Reunion Island es el lugar del mundo donde ocurren más ataques de tiburón en terminos relativos

### 2.1. Todos los ataques producidos en la India, ocurren a primera hora de la mañana

### 2.2. Las actividades con más riesgo de sufrir un ataque de tiburón son pescar o nadar

### 2.3. Reunion Island es el lugar del mundo donde ocurren más ataques de tiburón en terminos relativos